File: Code Samples/Queue - Abhijeet Chattar/JAVA/Dequeue.java
// Deque implementation in Java

class Deque {
    static final int MAX = 100;
    int arr[];
    int front;
    int rear;
    int size;

    public Deque(int size) {
        arr = new int[MAX];
        front = -1;
        rear = 0;
        this.size = size;
    }

    boolean isFull() {
        return ((front == 0 && rear == size - 1) || front == rear + 1);
    }

    boolean isEmpty() {
        return (front == -1);
    }

    void insertfront(int key) {
        if (isFull()) {
            System.out.println("Overflow");
            return;
        }

        if (front == -1) {
            front = 0;
            rear = 0;
        }

        else if (front == 0)
            front = size - 1;

        else
            front = front - 1;

        arr[front] = key;
    }

    void insertrear(int key) {
        if (isFull()) {
            System.out.println(" Overflow ");
            return;
        }

        if (front == -1) {
            front = 0;
            rear = 0;
        }

        else if (rear == size - 1)
            rear = 0;

        else
            rear = rear + 1;

        arr[rear] = key;
    }

    void deletefront() {
        if (isEmpty()) {
            System.out.println("Queue Underflow\n");
            return;
        }

        // Deque has only one element
        if (front == rear) {
            front = -1;
            rear = -1;
        } else if (front == size - 1)
            front = 0;

        else
            front = front + 1;
    }

    void deleterear() {
        if (isEmpty()) {
            System.out.println(" Underflow");
            return;
        }

        if (front == rear) {
            front = -1;
            rear = -1;
        } else if (rear == 0)
            rear = size - 1;
        else
            rear = rear - 1;
    }

    void display(){
        System.out.print("Items -> ");
        for(int i=front; i!=rear; i=(i+1)%size){
            System.out.print(arr[i] + " ");
        }
        System.out.print(arr[rear]);
    }
}

public class Dequeue {
    public static void main(String[] args) {

        Deque dq = new Deque(4);

        dq.insertrear(12);

        dq.insertrear(14);

        dq.deleterear();

        dq.insertfront(13);

        dq.insertfront(26);

        dq.display();

    }
}

File: Code Samples/Queue - Abhijeet Chattar/CPP/Dequeue.cpp
#include <iostream>
#define SIZE 10
using namespace std;

class Deque {
private:
    int arr[SIZE];
    int front;
    int rear;

public:
    Deque() {
        front = -1;
        rear = -1;
    }

    bool isEmpty() {
        if(front == -1)
            return true;
        return false;
    }

    bool isFull() {
        if((rear + 1) % SIZE == front)
            return true;
        return false;
    }

    void insertFront(int data) {
        if (isFull()) {
            cout << "Deque is full" << endl;
            return;
        }

        if (front == -1) {
            front = 0;
            rear = 0;
        } else if (front == 0) {
            front = SIZE - 1;
        } else {
            front--;
        }
        arr[front] = data;
    }

    void insertRear(int data) {
        if (isFull()) {
            cout << "Deque is full. Cannot insert at the rear." << endl;
            return;
        }

        if (front == -1) {
            front = 0;
            rear = 0;
        } else if (rear == SIZE - 1) {
            rear = 0;
        } else {
            rear++;
        }

        arr[rear] = data;
    }

    int deleteFront() {
        if (isEmpty()) {
            cout << "Deque is empty. Cannot delete from the front." << endl;
            return -1;
        }

        int data = arr[front];
        if (front == rear) {
            front = -1;
            rear = -1;
        } else if (front == SIZE - 1) {
            front = 0;
        } else {
            front++;
        }

        return data;
    }

    int deleteRear() {
        if (isEmpty()) {
            cout << "Deque is empty. Cannot delete from the rear." << endl;
            return -1;
        }

        int data = arr[rear];
        if (front == rear) {
            front = -1;
            rear = -1;
        } else if (rear == 0) {
            rear = SIZE - 1;
        } else {
            rear--;
        }

        return data;
    }

    void display() {
        if (isEmpty()) {
            cout << "Deque is empty." << endl;
            return;
        }

        int i = front;

        cout << "\nElements in dequeue -> ";
        for(int i=front; i!=rear; i=(i+1)%SIZE){
            cout << arr[i] << " ";
        }
        cout << arr[rear];
    }
};

int main() {
    Deque deque;

    deque.insertRear(1);
    deque.insertRear(2);
    deque.insertFront(0);
    deque.insertFront(-1);
    deque.deleteFront();
    deque.insertFront(-1);
    deque.display();
    return 0;
}
